packages/flutter_test/lib/src/finders.dart

```diff
@@ -460,181 +474,663 @@ class CommonFinders {
-    return collectAllElementsFrom(
-      WidgetsBinding.instance.rootElement!,
-      skipOffstage: skipOffstage,
+  Iterable<CandidateType> get allCandidates;
+
+  /// Returns a variant of this finder that only matches the first item
+  /// found by this finder.
+  FinderBase<CandidateType> get first => _FirstFinder<CandidateType>(this);
+
+  /// Returns a variant of this finder that only matches the last item
+  /// found by this finder.
+  FinderBase<CandidateType> get last => _LastFinder<CandidateType>(this);
+
+  /// Returns a variant of this finder that only matches the item at the
+  /// given index found by this finder.
+  FinderBase<CandidateType> at(int index) => _IndexFinder<CandidateType>(this, index);
+
+  /// Returns all the items in the given list that match this
+  /// finder's requirements.
+  ///
+  /// This is overridden to define the requirements of the finder when
+  /// implementing finders that directly extend [FinderBase]. If a finder can
+  /// be efficiently described just in terms of a predicate function, consider
+  /// mixing in [MatchFinderMixin] and implementing [MatchFinderMixin.matches]
+  /// instead.
+  @protected
+  Iterable<CandidateType> findInCandidates(Iterable<CandidateType> candidates);
+
+  /// Searches a set of candidates for those that meet the requirements set by
+  /// this finder and returns the result of that search.
+  ///
+  /// See also:
+  ///
+  ///   * [found] which will return the latest results without re-executing the
+  ///     search.
+  ///   * [tryEvaluate] which will indicate whether any results were found rather
+  ///     than directly returning results.
+  FinderResult<CandidateType> evaluate() {
+    if (!_cached || _found == null) {
+      _found = FinderResult<CandidateType>(describeMatch, findInCandidates(allCandidates));
+    }
+    return found;
+  }
+
+  /// Searches a set of candidates for those that meet the requirements set by
+  /// this finder and returns whether the search found any matching candidates.
+  ///
+  /// This is useful in cases where an action needs to be repeated while or
+  /// until a finder has results. The results from the search can be accessed
+  /// using the [found] property without re-executing the search.
+  ///
+  /// ## Sample code
+  ///
+  /// ```dart
+  /// testWidgets('Top text loads first', (WidgetTester tester) async {
+  ///   // Assume a widget is pumped with a top and bottom loading area, with
+  ///   // the texts "Top loaded" and "Bottom loaded" when loading is complete.
+  ///   // await tester.pumpWidget(...)
+  ///
+  ///   // Wait until at least one loaded widget is available
+  ///   WidgetFinder loadedFinder = find.textContaining('loaded');
+  ///   while (!loadedFinder.tryEvaluate()) {
+  ///     await tester.pump(const Duration(milliseconds: 100));
+  ///   }
+  ///
+  ///   expect(loadedFinder.found, hasLength(1));
+  ///   expect(tester.widget<Text>(loadedFinder.found).data, contains('Top'));
```

Comment:
From chat, does this actually work?

Reply:
Copied it over to a test and no, no it does not. I'll have to either fix this or (as we're discussing in chat) see if making it work is worth it. Thanks for adding a comment to keep track of it.

Reply:
Fixed.